Port the ros2_tracing page from ros-realtime documentation to primary documentation (#3621)
ros2/ros2_documentation: source/Tutorials/Advanced/ROS2-Tracing-Trace-and-Analyze.rst

How to use ros2_tracing to trace and analyze an application
===========================================================

This guide shows how to use `ros2_tracing <https://github.com/ros2/ros2_tracing>`_ to trace and analyze a ROS 2 application.
For this guide, the application will be `performance_test <https://gitlab.com/ApexAI/performance_test>`_.

Overview
--------

This guide covers:

1. running and tracing a ``performance_test`` run
2. analyzing the trace data using `tracetools_analysis <https://gitlab.com/ros-tracing/tracetools_analysis>`_ to plot the callback durations

Prerequisites
-------------

This guide is aimed at real-time systems.
See the :doc:`real-time system setup guide <../Miscellaneous/Building-Realtime-rt_preempt-kernel-for-ROS-2>`.
However, the guide will work if you are using a non-real-time system.

> .. note::
> 
>   This guide should work with all current/non-EOL distros of ROS 2.
> 

Reviewer: We could remove this note, since this page technically applies to a single ROS 2 distro. The fact that it works with that distro (and any other non-EOL distro for which documentation includes this tutorial) is kind of implied.

```suggestion
```

Installing and building
-----------------------

> Install ROS 2 by following the :doc:`installation instructions <../../Installation/Ubuntu-Install-Debians>`.

Reviewer: ```suggestion
Install ROS 2 by following the :doc:`installation instructions <../../Installation/Ubuntu-Install-Debians>`.

.. note::
  This guide should generally work with all supported Linux distributions.

```

.. code-block:: bash

   $ cd ~/
   $ mkdir -p tracing_ws/src
   $ cd tracing_ws/src/
   $ git clone https://gitlab.com/ApexAI/performance_test.git
   $ git clone https://gitlab.com/ros-tracing/tracetools_analysis.git
   $ cd ..

Install dependencies with rosdep.

.. code-block:: bash

   $ rosdep update
   $ rosdep install --from-paths src --ignore-src -y

Then build and configure ``performance_test`` for ROS 2.
See its `documentation <https://gitlab.com/ApexAI/performance_test/-/tree/master/performance_test#performance_test>`_.

.. code-block:: bash

   $ colcon build --packages-select performance_test --cmake-args -DPERFORMANCE_TEST_RCLCPP_ENABLED=ON

Next, we will run a ``performance_test`` experiment and trace it.

Tracing
-------

Step 1: Trace
^^^^^^^^^^^^^

  In one terminal, source the workspace and set up tracing.
  When running the command, a list of ROS 2 userspace events will be printed.
  It will also print the path to the directory that will contain the resulting trace (under ``~/.ros/tracing``).

  .. code-block:: bash

    $ # terminal 1
    $ cd ~/tracing_ws
    $ source install/setup.bash
    $ ros2 trace --session-name perf-test --list

>     

Reviewer: CI is complaining about the trailing whitespace here:

```suggestion

```

How to use ros2_tracing to trace and analyze an application
===========================================================

This guide shows how to use `ros2_tracing <https://github.com/ros2/ros2_tracing>`_ to trace and analyze a ROS 2 application.
For this guide, the application will be `performance_test <https://gitlab.com/ApexAI/performance_test>`_.

Overview
--------

This guide covers:

1. running and tracing a ``performance_test`` run
2. analyzing the trace data using `tracetools_analysis <https://github.com/ros-tracing/tracetools_analysis>`_ to plot the callback durations

Prerequisites
-------------

> This guide is aimed at real-time systems.

Reviewer 2: I think we should also mention here that this will only work on Linux.

Reviewer: Is this enough? e1006fe2c4c5e4ed7515d1b845a7e1eb5db2f1a2

Source the ROS 2 installation and verify that tracing is enabled:

.. code-block:: bash

   source /opt/ros/{DISTRO}/setup.bash
   ros2 run tracetools status

You should see ``Tracing enabled`` in the output.

Then create a workspace, and clone ``performance_test`` and ``tracetools_analysis``.

.. code-block:: bash

   cd ~/
   mkdir -p tracing_ws/src
   cd tracing_ws/src/
   git clone https://gitlab.com/ApexAI/performance_test.git
   git clone https://github.com/ros-tracing/tracetools_analysis.git
   cd ..

Install dependencies with rosdep.

.. code-block:: bash

   rosdep update
   rosdep install --from-paths src --ignore-src -y

Then build and configure ``performance_test`` for ROS 2.
See its `documentation <https://gitlab.com/ApexAI/performance_test/-/tree/master/performance_test#performance_test>`_.

.. code-block:: bash

   colcon build --packages-select performance_test --cmake-args -DPERFORMANCE_TEST_RCLCPP_ENABLED=ON

Next, we will run a ``performance_test`` experiment and trace it.

Tracing
-------

Step 1: Trace
^^^^^^^^^^^^^

  In one terminal, source the workspace and set up tracing.
  When running the command, a list of ROS 2 userspace events will be printed.

>   It will also print the path to the directory that will contain the resulting trace (under ``~/.ros/tracing``).

Reviewer 2: We don't typically indent our explanations, so let's just unindent this.  The same goes for all of the sections below.

The ``tracetools_analysis`` repository contains a `few sample notebooks <https://github.com/ros-tracing/tracetools_analysis/tree/{DISTRO}/tracetools_analysis/analysis>`_, including `one notebook to analyze subscription callback durations <https://github.com/ros-tracing/tracetools_analysis/blob/{DISTRO}/tracetools_analysis/analysis/callback_duration.ipynb>`_.

For this guide, we will plot the durations of the subscription callback in the subscriber node.

Install Jupyter notebook and bokeh, and then open the sample notebook.

.. code-block:: bash

   $ pip3 install bokeh
   $ jupyter notebook ~/tracing_ws/src/tracetools_analysis/tracetools_analysis/analysis/callback_duration.ipynb

This will open the notebook in the browser.

Replace the value for the ``path`` variable in the second cell to the path to the trace directory:

.. code-block:: python

   path = '~/.ros/tracing/perf-test'

Run the notebook by clicking the *Run* button for each cell.
Running the cell that does the trace processing might take a few minutes on the first run, but subsequent runs will be much quicker.

> You should get a plot that looks like this:

Reviewer 2: ```suggestion
You should get a plot that looks similar to this:
```

How to use ros2_tracing to trace and analyze an application
===========================================================

This guide shows how to use `ros2_tracing <https://github.com/ros2/ros2_tracing>`_ to trace and analyze a ROS 2 application.
For this guide, the application will be `performance_test <https://gitlab.com/ApexAI/performance_test>`_.

Overview
--------

This guide covers:

1. running and tracing a ``performance_test`` run
2. analyzing the trace data using `tracetools_analysis <https://gitlab.com/ros-tracing/tracetools_analysis>`_ to plot the callback durations

Prerequisites
-------------

> This guide is aimed at real-time systems. See the `real-time system setup guide 
> <Real-Time-Operating-System-Setup/Real-Time-Linux/rt_linux_index.md>`_.
> This guide is aimed at real-time systems.
> See the :doc:`real-time system setup guide <../Miscellaneous/Building-Realtime-rt_preempt-kernel-for-ROS-2>`.
> However, the guide will work if you are using a non-real-time system.
> 
>   This guide was written for ROS 2 Rolling on Ubuntu 22.04.
>   It should work on other ROS 2 distros or Ubuntu versions, but some things might need to be adjusted.

Reviewer: Something went wrong here:

```suggestion
This guide is aimed at real-time systems.
See the :doc:`real-time system setup guide <../Miscellaneous/Building-Realtime-rt_preempt-kernel-for-ROS-2>`.
However, the guide will work if you are using a non-real-time system.

.. note::

  This guide was written for ROS 2 Rolling on Ubuntu 22.04.
  It should work on other ROS 2 distros or Ubuntu versions, but some things might need to be adjusted.
```

Installing and building
-----------------------

Install ROS 2 by following the :doc:`installation instructions <../../Installation/Ubuntu-Install-Debians>`.
Install ``babeltrace`` and ``ros2trace``.

.. code-block:: bash

   $ sudo apt-get update
   $ sudo apt-get install -y babeltrace ros-{DISTRO}-ros2trace ros-{DISTRO}-tracetools-analysis

> Source the ROS 2 installation and verify that tracing is enabled:
> 
>    $ source /opt/ros/{DISTRO}/setup.bash
>    $ ros2 run tracetools status

Reviewer: Same here:

```suggestion
Source the ROS 2 installation and verify that tracing is enabled:

.. code-block:: bash

   $ source /opt/ros/{DISTRO}/setup.bash
   $ ros2 run tracetools status
```

   We simply create an experiment with a node publishing ~1 MB messages to another node as fast as possible for 60 seconds using the second highest real-time priority so that we don’t interfere with critical kernel threads.
   We need to run ``performance_test`` as ``root`` to be able to use real-time priorities.

   .. code-block:: bash

      $ # terminal 2
      $ sudo ./install/performance_test/lib/performance_test/perf_test -c rclcpp-single-threaded-executor -p 1 -s 1 -r 0 -m Array1m --reliability RELIABLE --max-runtime 60 --use-rt-prio 98

   If that last command doesn’t work for you (with an error like: “error while loading shared libraries”), run the slightly-different command below.
   This is because, for security reasons, we need to manually pass ``*PATH`` environment variables for some shared libraries to be found (see `this explanation <https://unix.stackexchange.com/a/251374>`_).

   .. code-block:: bash

      $ # terminal 2
      $ sudo env PATH="$PATH" LD_LIBRARY_PATH="$LD_LIBRARY_PATH" ./install/performance_test/lib/performance_test/perf_test -c rclcpp-single-threaded-executor -p 1 -s 1 -r 0 -m Array1m --reliability RELIABLE --max-runtime 60 --use-rt-prio 98

>   .. note::
>     If you're not using a real-time kernel, simply run:
> 
>     .. code-block:: bash
> 
>         $ # terminal 2
>         $ ./install/performance_test/lib/performance_test/perf_test -c rclcpp-single-threaded-executor -p 1 -s 1 -r 0 -m Array1m --reliability RELIABLE --max-runtime 60

Reviewer: Assuming that the indentation of these sub-sections is intentional (since the list was turned into sub-sections), the indentation of these lines needs to be fixed:

```suggestion
   .. note::

      If you're not using a real-time kernel, simply run:

      .. code-block:: bash

         $ # terminal 2
         $ ./install/performance_test/lib/performance_test/perf_test -c rclcpp-single-threaded-executor -p 1 -s 1 -r 0 -m Array1m --reliability RELIABLE --max-runtime 60
```

Author: The wrong lines were highlighted, but the text above it was fixed

How to use ros2_tracing to trace and analyze an application
===========================================================

This guide shows how to use `ros2_tracing <https://github.com/ros2/ros2_tracing>`_ to trace and analyze a ROS 2 application.
For this guide, the application will be `performance_test <https://gitlab.com/ApexAI/performance_test>`_.

Overview
--------

This guide covers:

1. running and tracing a ``performance_test`` run
2. analyzing the trace data using `tracetools_analysis <https://gitlab.com/ros-tracing/tracetools_analysis>`_ to plot the callback durations

Prerequisites
-------------

This guide is aimed at real-time systems.
See the :doc:`real-time system setup guide <../Miscellaneous/Building-Realtime-rt_preempt-kernel-for-ROS-2>`.
However, the guide will work if you are using a non-real-time system.

.. note::

  This guide was written for ROS 2 Rolling on Ubuntu 22.04.

>   It should work on other ROS 2 distros or Ubuntu versions, but some things might need to be adjusted.

Reviewer 2: I don't particularly think we should add this note here; it will be out-of-date when we backport it, and when we branch off the new distribution.  If there are things that should be different for the different distributions, we should fix that in the backport.

Reviewer: Sounds good to me. This guide should work with all current/non-EOL distros. @nightduck we just need to add `--branch {DISTRO}` to the 2 `git clone` commands.

Reviewer: > This guide should work with all current/non-EOL distros

Well, for Humble, we'll need to build from source, since the tracing instrumentation isn't included by default. Let's not backport to Humble for now, I'll do it manually as a follow-up PR.

Installing and building
-----------------------

> Install ROS 2 by following the :doc:`installation instructions <../../Installation/Ubuntu-Install-Debians>`.

Reviewer 2: Two things about this:

1.  We should mention somewhere that this will only work on Linux.
2.  Does this work on RHEL as well?  If so, we should just point to the generic Installation page.

Reviewer: > 1. We should mention somewhere that this will only work on Linux.

:+1: 

> 1. Does this work on RHEL as well? If so, we should just point to the generic Installation page.

Yes, it should work on RHEL (although I don't think this particular tutorial was ever tested on RHEL).

Author: Should we state that is has been tested on Ubuntu, but should work on Debian and RHEL?

Reviewer: I wouldn't necessarily mention that it has been tested on Ubuntu, but yeah we can say it should work on other Linux distros.

Install ``babeltrace`` and ``ros2trace``.

.. code-block:: bash

   $ sudo apt-get update
   $ sudo apt-get install -y babeltrace ros-{DISTRO}-ros2trace ros-{DISTRO}-tracetools-analysis

Source the ROS 2 installation and verify that tracing is enabled:

.. code-block:: bash

   $ source /opt/ros/{DISTRO}/setup.bash
   $ ros2 run tracetools status

You should see ``Tracing enabled`` in the output.

Then create a workspace, and clone ``performance_test`` and ``tracetools_analysis``.

.. code-block:: bash

   $ cd ~/
   $ mkdir -p tracing_ws/src
   $ cd tracing_ws/src/
   $ git clone https://gitlab.com/ApexAI/performance_test.git
   $ git clone https://gitlab.com/ros-tracing/tracetools_analysis.git
   $ cd ..

Install dependencies with rosdep.

.. code-block:: bash

   $ rosdep update

>    $ rosdep install --from-paths src --ignore-src -y --skip-keys "fastcdr rti-connext-dds-6.0.1 urdfdom_headers"

Reviewer 2: Are the `skip-keys` really necessary here?  I don't think either of the packages above directly depend on any of these packages, so I would think we wouldn't need it.  But I could be wrong.

Reviewer: Yeah, these should not be direct dependencies.

Then build and configure ``performance_test`` for ROS 2.
See its `documentation <https://gitlab.com/ApexAI/performance_test/-/tree/master/performance_test#performance_test>`_.

.. code-block:: bash

   $ colcon build --packages-select performance_test --cmake-args -DPERFORMANCE_TEST_RCLCPP_ENABLED=ON

Next, we will run a ``performance_test`` experiment and trace it.

Tracing
-------

**(Optional)** Step 0: Start an LTTng session daemon.
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

   For userspace tracing, the daemon does not need to be started as ``root``.
   Note that a non-root daemon will be spawned automatically by ``ros2 trace`` if it is not already running.

   .. code-block:: bash

>       $ lttng-sessiond --daemonize

Reviewer 2: I guess I'm a little confused by this section.  When and when would you not want the daemon started as root?  When do you need the daemon, and when don't you?

Reviewer: We always need the daemon to trace. We need a root daemon to trace the kernel, which we don't do in this tutorial. And, since a non-root daemon is spawned automatically when running `ros2 trace` if needed, we can just remove this step for simplicity.

Author: That's why I listed it as an optional step 0

Reviewer: Yeah, but then we might as well just not mention it.

Step 1: Trace
^^^^^^^^^^^^^

>    In one terminal, source the workspace and set up tracing.

Reviewer 2: For all of the lines in this section, they do not need to be indented 2 spaces.

Reviewer: I assumed this was a stylistic choice by @nightduck, but yeah they don't need to be indented.

Author: Twas a style choice. But this particular section is indented 3 spaces for some reason. Correcting.

  .. code-block:: bash

    $ # terminal 2
    $ sudo env PATH="$PATH" LD_LIBRARY_PATH="$LD_LIBRARY_PATH" ./install/performance_test/lib/performance_test/perf_test -c rclcpp-single-threaded-executor -p 1 -s 1 -r 0 -m Array1m --reliability RELIABLE --max-runtime 60 --use-rt-prio 98

  .. note::
    If you're not using a real-time kernel, simply run:

    .. code-block:: bash

        $ # terminal 2
        $ ./install/performance_test/lib/performance_test/perf_test -c rclcpp-single-threaded-executor -p 1 -s 1 -r 0 -m Array1m --reliability RELIABLE --max-runtime 60

Step 3: Validate Trace
^^^^^^^^^^^^^^^^^^^^^^

   Once the experiment is done, in the first terminal, press enter again to stop tracing.
   Use ``babeltrace`` to quickly look at the resulting trace.

   .. code-block:: bash

      $ babeltrace ~/.ros/tracing/perf-test | less

   The output of the above command is a human-readable version of the raw Common Trace Format (CTF) data, which is a list of trace events.
   Each event has a timestamp, an event type, some information on the process that generated the event, and the values of the fields of the given event type.

>    Use the arrow keys to scroll, or press q to exit.

Reviewer 2: ```suggestion
   Use the arrow keys to scroll, or press ``q`` to exit.
```
